xclim/indices/generic.py

```diff
@@ -352,6 +352,115 @@ def get_daily_events(
     return events
 
 
+@declare_relative_units(threshold="<data>")
+def spell_length_statistics(
+    data: xarray.DataArray,
+    threshold: Quantified,
+    window: int,
+    win_reducer: str,
+    op: str,
+    spell_reducer: str,
+    freq: str,
+    resample_before_rl: bool = True,
+    **indexer,
+):
+    """Statistics on spells lengths.
+
+    A spell is when a statistic (`win_reducer`) over a minimum number (`window`) of consecutive timesteps respects a condition (`op` `thresh`).
+    This returns a statistic over the spell's count or length.
+
+    Parameters
+    ----------
+    data: xr.DataArray
+        Input data.
+    threshold: Quantified
+        Threshold to test against.
+    window: int
+        Minimum length of a spell.
+    win_reducer: {'min', 'max', 'sum', 'mean'}
+        Reduction along the spell length to compute the spell value.
+        Note that this does not matter when `window` is 1.
+    op : {">", "gt", "<", "lt", ">=", "ge", "<=", "le", "==", "eq", "!=", "ne"}
+        Logical operator. Ex: spell_value > thresh.
+    spell_reducer: {'max', 'sum', 'count'}
+        Statistic on the spell lengths.
+    freq: str
+        Resampling frequency.
+    resample_before_rl: bool
+        Determines if the resampling should take place before or after the run
+        length encoding (or a similar algorithm) is applied to runs.
+    indexer:
+        Indexing parameters to compute the indicator on a temporal subset of the data.
+        It accepts the same arguments as :py:func:`xclim.indices.generic.select_time`.
+        Indexing is done after finding the days part of a spell, but before taking the spell statistics.
+
+    Examples
+    --------
+    >>> spell_statistics(
+    ...     tas,
+    ...     thresh="35 °C",
+    ...     window=7,
+    ...     op=">",
+    ...     win_reducer="min",
+    ...     spell_reducer="sum",
+    ...     freq="YS",
+    ... )
+
+    Here, a day is part of a spell if it is in any 7 day period where the minimum temperature is over 35°C. We then
+    return the annual sum of the spell lengths, so the total number of days in such spells.
+
+    >>> pram = rate2amount(pr)
+    >>> spell_statistics(
+    ...     pram,
+    ...     thresh="20 mm",
+    ...     window=5,
+    ...     op=">=",
+    ...     win_reducer="sum",
+    ...     spell_reducer="max",
+    ...     freq="YS",
+    ... )
+
+    Here, a day is part of a spell if it is in any 5 day period where the total accumulated precipitation reaches or exceeds
+    20 mm. We then return the length of the longest of such spells.
+    """
+    thresh = convert_units_to(
+        threshold,
+        data,
+        context=infer_context(standard_name=data.attrs.get("standard_name")),
+    )
+
+    if window == 1:  # Fast path
+        is_in_spell = compare(data, op, thresh)
+    else:
+        data_pad = data.pad(time=(0, window))
```

Comment:
I guess all this rolling is only necessary for `window_reducer == ="sum"`, yes? If you want to include a faster algorithm for other cases, a modification of `rle` should do the trick: 

```python 
from xclim.indices.run_length import _cumsum_reset_on_zero
mask = compare(data, op,thresh)
# find sequences of consecutive respected constraints
cs_s = _cumsum_reset_on_zero(mask)
# end of these sequences
cs_s = cs_s.where(mask.shift({"time": -1}, fill_value=0) == 0)
# propagate these end of sequences
is_in_spell = (cs_s.where(cs_s>=window).where(mask>0, 0).bfill("time")>0)
```

Reply:
Added this as-is in the generic function. Did you see a way to generalize this futher then the two cases I saw ? See https://github.com/Ouranosinc/xclim/blob/8b75acd1cdeb2e5b989353825ae0e6ec2466b8e9/xclim/indices/generic.py#L434-L436

Reply:
Even if it is weird, the "min" + "<" case is possible and I don't think it is feasible without both rolling. This fast path, if I understand it correctly, works when the condition on a the minimum window can only be true if it is true on each day of the window.

Reply:
What you say here is related to this comment, yes?:

> # For these two cases, a day can't be part of a spell if it doesn't respect the condition itself

or said otherwise, we must have "min" and ">"  / "max" and "<" correct? 

I agree that this fast-path is less flexible, but as you note it, it's more natural. I can see why the `op="sum"` could be interesting, but not the "min"&"<" frankly. 

Imagine you're trying to characterize wet spells, where the start condition is having 10 mm and say `window=3`.  

`pr = [0, 0, 20,20,20, 0,0] # in mm`

With the weird definition, the spell sequence is `[0,0,20,20,20]` because, why not haha?  You just end up adding `window-1` extra days no matter what, no?

Reply:
In agree that the (min, <) and (max, >) are weird, but I just don't want to forbid a user from using it solely on this judgement.